src/cargo/ops/cargo_add/mod.rs

```diff
@@ -0,0 +1,553 @@
+    /// Package to add dependencies to
+    pub spec: &'a crate::core::Package,
+    /// Dependencies to add or modify
+    pub dependencies: Vec<DepOp>,
+    /// Which dependency section to add these to
+    pub section: DepTable,
+    /// Act as if dependencies will be added
+    pub dry_run: bool,
+}
+
+/// Add dependencies to a manifest
+pub fn add(workspace: &crate::core::Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<()> {
+    let dep_table = options
+        .section
+        .to_table()
+        .into_iter()
+        .map(String::from)
+        .collect::<Vec<_>>();
+
+    let manifest_path = options.spec.manifest_path().to_path_buf();
+    let mut manifest = LocalManifest::try_new(&manifest_path)?;
+    let legacy = manifest.get_legacy_sections();
+    if !legacy.is_empty() {
+        anyhow::bail!(
+            "Deprecated dependency sections are unsupported: {}",
+            legacy.join(", ")
+        );
+    }
+
+    let mut registry = crate::core::registry::PackageRegistry::new(options.config)?;
+
+    let deps = {
+        let _lock = options.config.acquire_package_cache_lock()?;
+        registry.lock_patches();
+        options
+            .dependencies
+            .iter()
+            .map(|raw| {
+                resolve_dependency(
+                    &manifest,
+                    raw,
+                    workspace,
+                    &options.section,
+                    options.config,
+                    &mut registry,
+                )
+            })
+            .collect::<CargoResult<Vec<_>>>()?
+    };
+
+    let was_sorted = manifest
+        .get_table(&dep_table)
+        .map(TomlItem::as_table)
+        .map_or(true, |table_option| {
+            table_option.map_or(true, |table| is_sorted(table.iter().map(|(name, _)| name)))
+        });
+    for dep in deps {
+        if let Some(req_feats) = dep.features.as_ref() {
+            let req_feats: BTreeSet<_> = req_feats.iter().map(|s| s.as_str()).collect();
+
+            let available_features = dep
+                .available_features
+                .keys()
+                .map(|s| s.as_ref())
+                .collect::<BTreeSet<&str>>();
+
+            let mut unknown_features: Vec<&&str> =
+                req_feats.difference(&available_features).collect();
+            unknown_features.sort();
+
+            if !unknown_features.is_empty() {
+                options
+                    .config
+                    .shell()
+                    .warn(format!("unrecognized features: {unknown_features:?}"))?;
+            };
+        }
+
+        print_msg(&mut options.config.shell(), &dep, &dep_table)?;
+        if let Some(Source::Path(src)) = dep.source() {
+            if src.path == manifest.path.parent().unwrap_or_else(|| Path::new("")) {
+                anyhow::bail!(
+                    "cannot add `{}` as a dependency to itself",
+                    manifest.package_name()?
+                )
+            }
+        }
```

Comment:
Tried it out but instead I'm going to make the order
- print_msg
- check error
- feature warning

`print_msg` says `Adding cargo-list-test-fixture-dependency (local) to dev-dependencies.` and it feels less abrupt / friendlier to see that message before seeing an error about about (might be different if it was "Added" instead of "Adding")